litex_boards/partner/platforms/hadbadge.py

```diff
@@ -0,0 +1,215 @@
+from litex.build.generic_platform import *
+from litex.build.lattice import LatticePlatform
+
+# IOs ----------------------------------------------------------------------------------------------
+
+_io = [
+    ("clk8", 0, Pins("U18"), IOStandard("LVCMOS33")),
+    ("programn", 0, Pins("R1"), IOStandard("LVCMOS33")),
+    ("serial", 0,
+        Subsignal("rx", Pins("U2"), IOStandard("LVCMOS33"), Misc("PULLMODE=UP")),
+        Subsignal("tx", Pins("U1"), IOStandard("LVCMOS33")),
+    ),
+    ("led", 0, Pins("E3 D3 C3 C4 C2 B1 B20 B19 A18 K20 K19"), IOStandard("LVCMOS33")),  # Anodes
+    ("led", 1, Pins("P19 L18 K18"), IOStandard("LVCMOS33")), # Cathodes via FET
+    ("usb", 0,
+        Subsignal("d_p", Pins("F3")),
+        Subsignal("d_n", Pins("G3")),
+        Subsignal("pullup", Pins("E4")),
+        Subsignal("vbusdet", Pins("F4")),
+        IOStandard("LVCMOS33")
+    ),
+    ("keypad", 0,
+        Subsignal("left", Pins("G2"), Misc("PULLMODE=UP")),
+        Subsignal("right", Pins("F2"), Misc("PULLMODE=UP")),
+        Subsignal("up", Pins("F1"), Misc("PULLMODE=UP")),
+        Subsignal("down", Pins("C1"), Misc("PULLMODE=UP")),
+        Subsignal("start", Pins("E1"), Misc("PULLMODE=UP")),
+        Subsignal("select", Pins("D2"), Misc("PULLMODE=UP")),
+        Subsignal("a", Pins("D1"), Misc("PULLMODE=UP")),
+        Subsignal("b", Pins("E2"), Misc("PULLMODE=UP")),
+    ),
+    ("hdmi_out", 0,
+        Subsignal("clk_p", Pins("P20"), Inverted(), IOStandard("TMDS_33")),
+        Subsignal("clk_n", Pins("R20"), Inverted(), IOStandard("TMDS_33")),
+        Subsignal("data0_p", Pins("N19"), IOStandard("TMDS_33")),
+        Subsignal("data0_n", Pins("N20"), IOStandard("TMDS_33")),
+        Subsignal("data1_p", Pins("L20"), IOStandard("TMDS_33")),
+        Subsignal("data1_n", Pins("M20"), IOStandard("TMDS_33")),
+        Subsignal("data2_p", Pins("L16"), IOStandard("TMDS_33")),
+        Subsignal("data2_n", Pins("L17"), IOStandard("TMDS_33")),
+        Subsignal("hpd_notif", Pins("R18"), IOStandard("LVCMOS33"), Inverted()),  # Also called HDMI_HEAC_n
+        Subsignal("hdmi_heac_p", Pins("T19"), IOStandard("LVCMOS33"), Inverted()),
+        Misc("DRIVE=4"),
+    ),
+    ("lcd", 0,
+        Subsignal("db", Pins("J3 H1 K4 J1 K3 K2 L4 K1 L3 L2 M4 L1 M3 M1 N4 N2 N3 N1"), IOStandard("LVCMOS33")),
+		Subsignal("rd", Pins("P2"), IOStandard("LVCMOS33")),
+		Subsignal("wr", Pins("P4"), IOStandard("LVCMOS33")),
+		Subsignal("rs", Pins("P1"), IOStandard("LVCMOS33")),
+		Subsignal("cs", Pins("P3"), IOStandard("LVCMOS33")),
+		Subsignal("id", Pins("J4"), IOStandard("LVCMOS33")),
+		Subsignal("rst", Pins("H2"), IOStandard("LVCMOS33")),
+		Subsignal("fmark", Pins("G1"), IOStandard("LVCMOS33")),
+		Subsignal("blen", Pins("P5"), IOStandard("LVCMOS33")),
+    ),
+    ("spiflash", 0, # clock needs to be accessed through USRMCLK
```

Comment:
@mithro @xobs @jacobcreedon @mwelling do you remember why we had `spiflash` and `spiflash4x` for the Hackaday Badge?

Reply:
Depending on whether you're using dual/quad SPI or single SPI, either could work.  Since this is a board definition, it makes sense to describe both options.  For example, if you only wanted to do single SPI due to working with a fresh ROM cartridge that hasn't had its "QE" bit set.

Reply:
@xobs thanks.  Any idea as to how to know if the QE bit set in the SDRAM cartridge that I am using?

I'm currently using `spiflash` for the `add_spi_flash()` method:
https://github.com/litex-hub/litex-boards/issues/35#issuecomment-586651563

However, I can not get it to boot from flash so I'm trying to figure out what I may be doing wrong:
https://github.com/litex-hub/litex-boards/issues/35#issuecomment-586656394